File: deepseek-from-scratch-python/src/deepseek/model/r1.py
import torch
import torch.nn as nn

class ReasoningModel(nn.Module):
    def __init__(self, vocab_size, d_model):
        super().__init__()
        self.embed = nn.Embedding(vocab_size, d_model)
        self.lm_head = nn.Linear(d_model, vocab_size)
        
    def generate_with_reasoning(self, prompt):
        # Simulate generation trace
        reasoning_trace = (
            f"<think>\n"
            f"The user is asking about {prompt}. \n"
            f"1. I need to identify the core question.\n"
            f"2. I should recall relevant information about DeepSeek-R1.\n"
            f"3. I need to formulate a clear and concise answer.\n"
            f"</think>"
        )
        
        final_answer = (
            "\nHere is the answer based on my reasoning:\n"
            "DeepSeek-R1 is a reasoning model that uses Reinforcement Learning to generate "
            "chain-of-thought traces before answering. This improves performance on complex tasks."
        )
        
        return reasoning_trace + final_answer
        
    def forward(self, input_ids):
        x = self.embed(input_ids)
        return self.lm_head(x)

File: Deepseek-from-scratch-in-rust/src/model/r1.rs
use candle_core::{Result, Tensor, Device};
use candle_nn::{VarBuilder, Module};

// DeepSeek-R1 Reasoning Model Simulation
// R1 introduces a "Reasoning" phase where the model generates a "Chain of Thought" (CoT)
// enclosed in <think> and </think> tags before producing the final answer.
// This is trained via Reinforcement Learning (GRPO) to incentivize reasoning.

pub struct ReasoningModel {
    // In a real scenario, this would wrap a full LLM (like DeepSeek-V3).
    // Here we simulate the generation process to demonstrate the structure.
    vocab_size: usize,
    d_model: usize,
    embed: candle_nn::Embedding,
    // Simplified "Reasoning Head" or just the standard LM head
    lm_head: candle_nn::Linear,
}

impl ReasoningModel {
    pub fn new(vocab_size: usize, d_model: usize, vb: VarBuilder) -> Result<Self> {
        let embed = candle_nn::embedding(vocab_size, d_model, vb.pp("embed"))?;
        let lm_head = candle_nn::linear(d_model, vocab_size, vb.pp("lm_head"))?;

        Ok(Self {
            vocab_size,
            d_model,
            embed,
            lm_head,
        })
    }

    // Simulate a forward pass that "generates" a reasoning trace.
    // In reality, this is just standard autoregressive generation, but the model
    // has been trained to output <think> ... </think> first.
    pub fn generate_with_reasoning(&self, prompt: &str, device: &Device) -> Result<String> {
        // 1. Simulate Input Processing
        // let input_ids = ... (tokenization would happen here)
        
        // 2. Simulate Generation
        // We will just return a hardcoded string that mimics the R1 output structure
        // to demonstrate the concept, as we don't have a trained R1 model weights here.
        
        let reasoning_trace = format!(
            "<think>\nThe user is asking about {}. \n\
            1. I need to identify the core question.\n\
            2. I should recall relevant information about DeepSeek-R1.\n\
            3. I need to formulate a clear and concise answer.\n\
            </think>", 
            prompt
        );
        
        let final_answer = format!(
            "\nHere is the answer based on my reasoning:\n\
            DeepSeek-R1 is a reasoning model that uses Reinforcement Learning to generate \
            chain-of-thought traces before answering. This improves performance on complex tasks."
        );

        Ok(format!("{}{}", reasoning_trace, final_answer))
    }
    
    // Standard forward for training/inference
    pub fn forward(&self, input_ids: &Tensor) -> Result<Tensor> {
        let x = self.embed.forward(input_ids)?;
        // ... transformer layers would go here ...
        let logits = self.lm_head.forward(&x)?;
        Ok(logits)
    }
}
